pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -318,21 +368,74 @@ func (rd *restoreDataProcessor) openSSTs(
 		return nil
 	}
 
+	hashSpan := func(span roachpb.Span) (uint32, error) {
+		hash := fnv.New32a()
+		_, err := hash.Write(span.Key)
+		if err != nil {
+			return 0, err
+		}
+
+		return hash.Sum32() % uint32(rd.numWorkers), nil
+	}
+
 	log.VEventf(ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
 
-	storeFiles := make([]storageccl.StoreFile, 0, len(EntryFiles{}))
-	for _, file := range entry.Files {
+	storeFiles := make([]storageccl.StoreFile, 0, len(entry.Files))
+	var sstOverheadBytesPerFile int64
+	if rd.spec.Encryption != nil {
+		sstOverheadBytesPerFile = sstReaderEncryptedOverheadBytesPerFile
+	} else {
+		sstOverheadBytesPerFile = sstReaderOverheadBytesPerFile
+	}
+
+	for idx := 0; idx < len(entry.Files); {
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		if err := rd.mem.Grow(ctx, sstOverheadBytesPerFile); err != nil {
+			// If we failed to allocate more memory, send the iterator
+			// containing the files we have right now.
+			if len(storeFiles) > 0 {
+				iterOpts := storage.IterOptions{
+					RangeKeyMaskingBelow: rd.spec.RestoreTime,
+					KeyTypes:             storage.IterKeyTypePointsAndRanges,
+					LowerBound:           keys.LocalMax,
+					UpperBound:           keys.MaxKey,
+				}
+				iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, rd.spec.Encryption, iterOpts)
+				if err != nil {
+					return err
+				}
+				workerNum, err := hashSpan(entry.Span)
+				if err != nil {
+					return err
+				}
+
+				log.VInfof(ctx, 2, "sending iterator after %d out of %d files due to insufficient memory", idx, len(entry.Files))
+				if err := sendIter(iter, dirs, workerSSTChannels[workerNum], int64(len(storeFiles))*sstReaderOverheadBytesPerFile, false); err != nil {
+					return err
+				}
+				storeFiles = storeFiles[:0]
+			}
+
+			select {
+			case <-time.After(sstReaderMemWaitSec * time.Second):
```

Comment:
can we do better than sleeping here?

What if we had a return channel or quotapool, so that when `cleanup` runs and shrinks, it unblocks the opener?

Reply:
Done, added a return channel.